Added icons to hide and show ref nodes of the graph

ericf-axosoft opened the pull request:
# Summary of changes
- Added icons (`eye-slash` and `eye`) needed to show/hide branches.
- Used the graph input property `hiddenRefsById` to be able to show/hide `refs` by using an `id`.
- Used the graph input property `onToggleRefVisibilityClick` event to be able to save `ref` nodes visibility in Gitlens.
- Added code to store hidden `refs`.

![Captura de pantalla 2022-09-28 a las 14 18 54](https://user-images.githubusercontent.com/90025366/192776966-07bfecab-818d-48b0-9b56-98bad759209b.png)

<img width="971" alt="Captura de pantalla 2022-09-30 a las 15 25 03" src="https://user-images.githubusercontent.com/90025366/193279658-28916e4d-5306-432d-b5f2-75ce03c37846.png">

https://user-images.githubusercontent.com/90025366/193278942-babbf1f8-389c-424f-bf52-ff7ef8857ed6.mov

# Notes
- I have created this PR as draft as we still need more design.
- This PR works with: https://github.com/gitkraken/GitKrakenComponents/pull/65

# Task
https://github.com/gitkraken/GitKrakenComponents/issues/61

ericf-axosoft commented:
@axosoft-ramint @eamodio we still need to discuss how to add the context menu with the hidden branches to be able to show again them in Gitlens side (when clicking the branch column button proposed by Samuel/Justin. See proposed solution in comment: https://github.com/gitkraken/GitKrakenComponents/issues/61#issuecomment-1255135530). 

For now, I have exposed the event, but maybe we will need to add extra DOM context properties in the graph component.

eamodio commented:
![Quality gate](https://app.embold.io/assets/images/svg/shield-off.svg) Quality gate: **_Failed_** 
 


            
 #### 8 Issues 
* ![Code issue](https://app.embold.io/assets/images/svg/code-issue.svg) 8 Code issues 

 [Details](https://app.embold.io/organization/gh/gitkraken/repositories/e23a25993521c5b78d5e00d144892422/pull-request/2211)

ericf-axosoft commented:
@axosoft-ramint I have just submitted the changes related to the context menu for the hidden refs. You will have to update the code of both PRs in your local to be able to test it (Graph component and Gitlens side).

axosoft-ramint reviewed (changes requested):
When I hide enough branches with long names (3 or more in my case), the text on the popup menu goes off my screen:

![image](https://user-images.githubusercontent.com/67011668/193502447-2e9d91f5-9355-46bf-ba85-ae813ab13732.png)

Also, I recall we talked about removing the count next to the icon, but I'll let @eamodio confirm that one.

ericf-axosoft commented:
@axosoft-ramint @eamodio I have modified Gitlens and graph component side to address following changes and fixes that we commented:
- Remove count of hidden refs (of the branch column button to show hidden refs).
- Use branch / remote / tag icons instead of the eye in the list of hidden refs.
- Change the hide icon of the ref to the right instead of the left position.
- Add an action to the context menu (right click on the ref) to hide the branch.
- CSS style issue regarding the hidden refs menu when branches have long names.
- Fixed a display issue when showing again some branches: sometimes branches where displayed in a wrong position. 

See recording for more details:

https://user-images.githubusercontent.com/90025366/193653176-08638379-0b53-4f41-ad8b-29aee4981c07.mov

axosoft-ramint reviewed (commented):
A few minor annoyances with the new UX. If we want to handle these in follow-up work I'm fine with it, since it's mostly functional.

1. With the hide button on the right side, if the ref label expands outside of the ref column width, you cannot reach the hide button sometimes (the label contracts when trying):

![cant get to the button](https://user-images.githubusercontent.com/67011668/193736949-3750828c-bd2f-47b0-ab84-364388431042.gif)

2. I expected to be able to toggle the menu by clicking the button. It's quite annoying that clicking it again does not hide the menu and I have to click elsewhere for it to go away:

![cant toggle](https://user-images.githubusercontent.com/67011668/193736963-709eba08-4dec-413f-8c93-95047f745978.gif)

If we want to address these in follow-up work, we can merge this. Let me know what you think.

ericf-axosoft commented:
> A few minor annoyances with the new UX. If we want to handle these in follow-up work I'm fine with it, since it's mostly functional.
> 
> 1. With the hide button on the right side, if the ref label expands outside of the ref column width, you cannot reach the hide button sometimes (the label contracts when trying):
> 
> ![cant get to the button](https://user-images.githubusercontent.com/67011668/193736949-3750828c-bd2f-47b0-ab84-364388431042.gif) [ ![cant get to the button](https://user-images.githubusercontent.com/67011668/193736949-3750828c-bd2f-47b0-ab84-364388431042.gif) ](https://user-images.githubusercontent.com/67011668/193736949-3750828c-bd2f-47b0-ab84-364388431042.gif) [ ](https://user-images.githubusercontent.com/67011668/193736949-3750828c-bd2f-47b0-ab84-364388431042.gif)
> 
> 2. I expected to be able to toggle the menu by clicking the button. It's quite annoying that clicking it again does not hide the menu and I have to click elsewhere for it to go away:
> 
> ![cant toggle](https://user-images.githubusercontent.com/67011668/193736963-709eba08-4dec-413f-8c93-95047f745978.gif) [ ![cant toggle](https://user-images.githubusercontent.com/67011668/193736963-709eba08-4dec-413f-8c93-95047f745978.gif) ](https://user-images.githubusercontent.com/67011668/193736963-709eba08-4dec-413f-8c93-95047f745978.gif) [ ](https://user-images.githubusercontent.com/67011668/193736963-709eba08-4dec-413f-8c93-95047f745978.gif)
> 
> If we want to address these in follow-up work, we can merge this. Let me know what you think.

The first issue was already commented by Eric A. on Slack on Friday. I will create a separate task to fix it. See https://gitkraken-hq.slack.com/archives/C02L4R4N4QG/p1664547693801509?thread_ts=1664545944.103479&cid=C02L4R4N4QG

As for the second issue, I will take a look at it just like the others minor styling issues that commented yesterday Eric A.: https://gitkraken-hq.slack.com/archives/C02L4R4N4QG/p1664823358523769?thread_ts=1664545944.103479&cid=C02L4R4N4QG

axosoft-ramint reviewed (approved):
Approving with the understanding that we can address UX issues separately.

ericf-axosoft commented:
@axosoft-ramint fixed case 2 in Graph component side in commit: [Fixed: able to toggle the menu by clicking the button](https://github.com/gitkraken/GitKrakenComponents/pull/65/commits/2a9b78498e788f864efbfc8c366630d1207a8504) 

Just created following tasks to address the minor styling issues of hidden refs:
https://github.com/gitkraken/GitKrakenComponents/issues/98
https://github.com/gitkraken/GitKrakenComponents/issues/99

eamodio commented:
@ericf-axosoft FYI, I'm still digging into this -- I started out just making some changes/fixing some issues to get this in, but I still have questions and need to dig deeper. Will post more later.

eamodio reviewed (changes requested):
I checked in a set of changes, but I left some inline comments that need to still be addressed.

Outside of that the remaining issue is dealing with this question I raised on the component PR:
>A UX question -- when using the eye icon to hide the ref, I kind of expected it to hide the whole ref, if there was both a local and remote(s) at the same point. But right now I need hide the first (which is local), then hide each remote. While with the context menus it feels a "little" different since I could target each branch specifically (e.g. clicking on the icons individually), but if I click on the overall ref (not directly on a specific icon) it still feels strange not to hide the whole thing.
>
>I think both the eye and the right-click on the overall ref (not on a specific icon) should hide the whole ref. And in the context menu we could have a different menu for that case -- "Hide from Graph", vs the more specific "Hide Branch" (or "Hide Branch from Graph"

As if we change the eventing we'll need corresponding updates here.

ericf-axosoft commented on src/storage.ts:
```diff
@@ -217,6 +218,13 @@ export interface StoredGraphColumn {
 	width?: number;
 }
 
+export interface StoredGraphHiddenRef {
+	id: string;
+	name: string;
+	type: string;
+	avatarUrl?: string;
+}
```
Same response as: https://github.com/gitkraken/GitKrakenComponents/pull/65#issuecomment-1268031596

ericf-axosoft commented on src/git/models/reference.ts:
```diff
@@ -101,15 +101,18 @@ export namespace GitRevision {
 	readonly remote: boolean;
 	readonly upstream?: { name: string; missing: boolean };
 	repoPath: string;
+	avatarUrl?: string;
```
Same response as: https://github.com/gitkraken/GitKrakenComponents/pull/65#issuecomment-1268031596

We need `avatarUrl` on the VSCode DOM context to be able to set the `hiddenRefsById` when using Gitlens context menu to hide a branch.

eamodio replied:
Then I think we should create a "graph only" wrapper for the `GitBranchReference` rather than adding it to the general model. Because `GitBranch` itself doesn't have an `avatarUrl` (and I'm not convinced that it should at this point)

ericf-axosoft commented on src/plus/webviews/graph/graphWebview.ts:
```diff
@@ -940,6 +969,25 @@ export class GraphWebview extends WebviewBase<State> {
+			};
+		}
+
+		return hiddenRefsById;
```
Ups, I forgot to remove that mapping after refactoring. 

This is fixed in your commit: `Changes from code review`.

ericf-axosoft commented on src/plus/webviews/graph/graphWebview.ts:
```diff
@@ -1116,6 +1166,15 @@ export class GraphWebview extends WebviewBase<State> {
+		let hiddenRefs = this.container.storage.getWorkspace('graph:hiddenRefs');
+		hiddenRefs = visible
+			? deleteRecordValue(hiddenRefs, ref.id)
+			: updateRecordValue(hiddenRefs, ref.id, ref);
```
This is fixed in your commit: `Changes from code review`.

ericf-axosoft commented on src/webviews/apps/plus/graph/graph.tsx:
```diff
@@ -80,6 +82,10 @@ export class GraphApp extends App<State> {
+                    onHiddenRefChange={debounce<GraphApp['onHiddenRefChanged']>(
+						(ref: GraphHiddenRef, visible: boolean) => this.onHiddenRefChanged(ref, visible),
+						250,
+					)}
```
Done in commit: `Removed unnecessary debounce`

ericf-axosoft commented on src/webviews/apps/plus/graph/graph.tsx:
```diff
@@ -141,6 +147,13 @@ export class GraphApp extends App<State> {
 
+            case DidChangeHiddenRefsNotificationType.method:
+				onIpc(DidChangeHiddenRefsNotificationType, msg, (params, type) => {
+					const newState = { ...this.state, hiddenRefs: params.hiddenRefs };
```
This is fixed in your commit: `Changes from code review`.

ericf-axosoft commented on src/git/models/branch.ts:
```diff
@@ -39,6 +39,10 @@ export interface BranchSortOptions {
 
 export class GitBranch implements GitBranchReference {
+	static getId(repoPath: string, remote: boolean, name: string): string {
+		return `${repoPath}|${remote ? 'remotes/' : 'heads/'}${name}`;
```
This is fixed in your commit: `Changes from code review`.

ericf-axosoft commented on src/plus/webviews/graph/protocol.ts:
```diff
@@ -38,6 +40,7 @@ export interface State {
 	trialBanner?: boolean;
 	workingTreeStats?: GraphWorkingTreeStats;
 	searchResults?: DidSearchParams['results'];
+	hiddenRefs?: HiddenRefsById;
```
This is fixed in your commit: `Changes from code review`.

d13 commented on src/plus/webviews/graph/graphWebview.ts:
```diff
@@ -1066,6 +1097,41 @@ export class GraphWebview extends WebviewBase<State> {
+		return this.container.storage.getWorkspace('graph:hiddenRefs');
+	}
+
+	private async getHiddenRefsById(
+		hiddenRefs: Record<string, GraphHiddenRef> | undefined
+	): Promise<GraphHiddenRefs | undefined> {
+		if (hiddenRefs == null) return undefined;
+
+		if (this.repository == null) {
+			this.repository = this.container.git.getBestRepositoryOrFirst();
+			if (this.repository == null) return undefined;
+		}
+
+		const [hiddenBranches, hiddenTags] = await Promise.all([
+			this.repository.getBranches({
+				filter: b => !b.current && hiddenRefs[b.id] != undefined,
+			}),
+			this.repository.getTags({
+				filter: t => hiddenRefs[t.id] != undefined,
+			}),
+		]);
```
This validation has too performance impact. I would ask we not ship v13 with this validation and have it as tech debt to solve after we launch.

d13 replied:
@eamodio is out of the office. We discussed this yesterday and we're in agreement.

ericf-axosoft replied:
Modified in commit: `Commented hidden refs validation for v13, as it has too much performance impact. Will be solved after we launch v13 as tech debt`.

eamodio commented:
@ericf-axosoft please make sure to rebase this and clean up (squash) any unnecessary commits. From a very quick look there will definitely be some conflicts with main

ericf-axosoft commented:
> @ericf-axosoft please make sure to rebase this and clean up (squash) any unnecessary commits. From a very quick look there will definitely be some conflicts with main

@eamodio @d13 @axosoft-ramint I just resolved the conflicts. Please, take a look at this and let me know if it is ok so we can start merging both PRs (Graph component and Gitlens side changes). Thanks.

d13 commented on src/plus/webviews/graph/graphWebview.ts:
```diff
@@ -1087,6 +1119,53 @@ export class GraphWebview extends WebviewBase<State> {
+		// return filteredHiddenRefsById;
+
+		// For v13, we return directly the hidden refs without validating them
+		const hiddenRefsById: GraphHiddenRefs = {};
+		for (const [id, graphHiddenRef] of Object.entries(hiddenRefs)) {
+			hiddenRefsById[id] = {
+				...graphHiddenRef,
+			};
+		}
```
Looks like this recreates the same object being passed-in. Is this work needed?

ericf-axosoft replied:
Changed in commit: `Removed logic that recreates the same object`

d13 commented:
Can some of these commits be cleaned-up / squashed together? Otherwise LGTM

axosoft-ramint reviewed (changes requested):
Two concerns when testing today:

1. When I hide a branch, and then rename that branch, the branch reappears in the graph with the new name (which is fine, though not ideal), but the old name of the branch is still in the "hidden refs" list, and toggling it just makes it go away. Ideally, the branch would not reappear, and the branch name in the hidden list would be updated. Not critical enough to hold back this feature entirely but definitely feels a bit strange.
2. When I have multiple repos open in the graph, it is not clear which repo a hidden ref belongs to. For example, I have three repos open and hide `development`, `staging` and `master` on them, and see this:

![image](https://user-images.githubusercontent.com/67011668/194972361-0617def3-ff59-4ff2-98ba-88f9786f0265.png)

Might need to append repo name to the branch in that list, or even better, have it filter to only show the hidden refs in the currently-opened repo.

ericf-axosoft commented:
> Two concerns when testing today:
> 
> 1. When I hide a branch, and then rename that branch, the branch reappears in the graph with the new name (which is fine, though not ideal), but the old name of the branch is still in the "hidden refs" list, and toggling it just makes it go away. Ideally, the branch would not reappear, and the branch name in the hidden list would be updated. Not critical enough to hold back this feature entirely but definitely feels a bit strange.
> 2. When I have multiple repos open in the graph, it is not clear which repo a hidden ref belongs to. For example, I have three repos open and hide `development`, `staging` and `master` on them, and see this:
> 
> ![image](https://user-images.githubusercontent.com/67011668/194972361-0617def3-ff59-4ff2-98ba-88f9786f0265.png)
> 
> Might need to append repo name to the branch in that list, or even better, have it filter to only show the hidden refs in the currently-opened repo.

@axosoft-ramint the filtering was already done to avoid those things you commented, but we decided not to ship in v13 as it has too much performance impact. We will have it as tech debt to solve after we launch.   

See comment: https://github.com/gitkraken/vscode-gitlens/pull/2211#discussion_r990117432

axosoft-ramint reviewed (approved):
Approved with the understanding that we will address the above issues later.

ericf-axosoft commented:
> Approved with the understanding that we will address the above issues later.

I have just created following issue to address that in the next iteration: https://github.com/gitkraken/vscode-gitlens/issues/2262
